src/mul.jl

```diff
@@ -362,6 +362,25 @@ end
 *(A::UpperOrLowerTriangular{<:Any,<:LayoutMatrix}, B::UpperOrLowerTriangular{<:Any,<:AdjOrTrans{<:Any,<:LayoutMatrix}}) = mul(A, B)
 *(A::UpperOrLowerTriangular{<:Any,<:AdjOrTrans{<:Any,<:LayoutMatrix}}, B::UpperOrLowerTriangular{<:Any,<:AdjOrTrans{<:Any,<:LayoutMatrix}}) = mul(A, B)
 
+*(A::SymTridiagonal{<:Any,<:LayoutVector}, B::SymTridiagonal{<:Any,<:LayoutVector}) = mul(A, B)
+*(A::SymTridiagonal{<:Any,<:LayoutVector}, B::Tridiagonal{<:Any,<:LayoutVector}) = mul(A, B)
+*(A::SymTridiagonal{<:Any,<:LayoutVector}, B::Bidiagonal{<:Any,<:LayoutVector}) = mul(A, B)
+*(A::SymTridiagonal{<:Any,<:LayoutVector}, B::UpperOrLowerTriangular{<:Any,<:LayoutMatrix}) = mul(A, B)
+*(A::SymTridiagonal, B::Diagonal{<:Any,<:LayoutVector}) = mul(A, B) # ambiguity
+*(A::Tridiagonal{<:Any,<:LayoutVector}, B::Tridiagonal{<:Any,<:LayoutVector}) = mul(A, B)
+*(A::Tridiagonal{<:Any,<:LayoutVector}, B::SymTridiagonal{<:Any,<:LayoutVector}) = mul(A, B)
+*(A::Tridiagonal{<:Any,<:LayoutVector}, B::Bidiagonal{<:Any,<:LayoutVector}) = mul(A, B)
+*(A::Tridiagonal{<:Any,<:LayoutVector}, B::UpperOrLowerTriangular{<:Any,<:LayoutMatrix}) = mul(A, B)
+*(A::Bidiagonal{<:Any,<:LayoutVector}, B::Bidiagonal{<:Any,<:LayoutVector}) = mul(A, B)
+*(A::Bidiagonal{<:Any,<:LayoutVector}, B::SymTridiagonal{<:Any,<:LayoutVector}) = mul(A, B)
+*(A::Bidiagonal{<:Any,<:LayoutVector}, B::Tridiagonal{<:Any,<:LayoutVector}) = mul(A, B)
+*(A::Bidiagonal{<:Any,<:LayoutVector}, B::UpperOrLowerTriangular{<:Any,<:LayoutMatrix}) = mul(A, B)
+*(A::UpperOrLowerTriangular{<:Any,<:LayoutMatrix}, B::SymTridiagonal{<:Any,<:LayoutVector}) = mul(A, B)
+*(A::UpperOrLowerTriangular{<:Any,<:LayoutMatrix}, B::Tridiagonal{<:Any,<:LayoutVector}) = mul(A, B)
+*(A::UpperOrLowerTriangular{<:Any,<:LayoutMatrix}, B::Bidiagonal{<:Any,<:LayoutVector}) = mul(A, B)
+*(A::UpperOrLowerTriangular{<:Any,<:LayoutMatrix}, B::Diagonal{<:Any,<:LayoutVector}) = mul(A, B) # ambiguity
+*(A::Diagonal{<:Any,<:LayoutVector}, B::SymTridiagonal{<:Any,<:LayoutVector}) = mul(A, B)
+*(A::Diagonal{<:Any,<:LayoutVector}, B::UpperOrLowerTriangular{<:Any,<:LayoutMatrix}) = mul(A, B)
```

Comment:
This should forward to `*(A::Diagonal, B.data::LayoutMatrix)`, so do we need to specialize the method for the triangular wrapper? Or will specializing it for `LayoutMatrix` suffice?

Reply:
Are you referring to all four of those methods? I think actually the second and fourth methods in your highlight aren't needed, so I could remove those. Otherwise I'm not sure what you mean

Reply:
Actually, they are all needed due to ambiguities that these fixes have to introduce

Reply:
I'm referring to the `*(A::Diagonal{<:Any,<:LayoutVector}, B::UpperOrLowerTriangular{<:Any,<:LayoutMatrix})` method on line 383. Does this lead to ambiguities? If it does, I think this should be resolved in `LinearAlgebra` as well.

Reply:
This is the complete error from the ambiguity that made me add that line (the other direction doesn't have an ambiguity)

```julia
julia> InfDiagonal() * InfUnitUpperTriangular() # has to be unit to get the ambiguity
ERROR: MethodError: *(::Diagonal{Float64, Main.InfiniteArrays.InfVec{Xoshiro}}, ::UnitUpperTriangular{Float64, Main.InfiniteArrays.InfMat{Xoshiro}}) is ambiguous.

Candidates:
  *(D::Diagonal, A::UnitUpperTriangular)
    @ LinearAlgebra C:\Users\djv23\.julia\juliaup\julia-1.10.4+0.x64.w64.mingw32\share\julia\stdlib\v1.10\LinearAlgebra\src\diagonal.jl:578       
  *(A::Diagonal{<:Any, <:LayoutVector}, B::AbstractMatrix)
    @ ArrayLayouts c:\Users\djv23\.julia\dev\ArrayLayouts.jl\src\ArrayLayouts.jl:220
  *(D::Diagonal, A::LinearAlgebra.AbstractTriangular)
    @ LinearAlgebra C:\Users\djv23\.julia\juliaup\julia-1.10.4+0.x64.w64.mingw32\share\julia\stdlib\v1.10\LinearAlgebra\src\diagonal.jl:835       
  *(A::AbstractMatrix, B::LinearAlgebra.AbstractTriangular)
    @ LinearAlgebra C:\Users\djv23\.julia\juliaup\julia-1.10.4+0.x64.w64.mingw32\share\julia\stdlib\v1.10\LinearAlgebra\src\triangular.jl:1500    
  *(D::Diagonal, A::AbstractMatrix)
    @ LinearAlgebra C:\Users\djv23\.julia\juliaup\julia-1.10.4+0.x64.w64.mingw32\share\julia\stdlib\v1.10\LinearAlgebra\src\diagonal.jl:292       

Possible fix, define
  *(::Diagonal{T, <:LayoutVector{T}} where T, ::UnitUpperTriangular)

Stacktrace:
 [1] top-level scope
   @ REPL[16]:1
```

I could make the method be `Union{UnitUpperTriangular, UnitLowerTriangular}` instead.

Reply:
> I could make the method be Union{UnitUpperTriangular, UnitLowerTriangular} instead. 

Actually nevermind, that just adds another ambiguity that forces me to add the more general line.. now I remember.

Reply:
Could you provide a MWE? How are `InfUnitUpperTriangular` and `InfDiagonal` defined? Ideally, `LinearAlgebra` should not be introducing these ambiguities, but we may have to add these methods for now, and version-limit these later.

Reply:
They are defined in this PR, built from my `InfVec` and `InfMat` definitions. Without this PR, you can reproduce it using this MWE:

```julia
using ArrayLayouts, LinearAlgebra
struct MyVec <: LayoutVector{Float64} 
    A::Vector{Float64}
end
struct MyMat <: LayoutMatrix{Float64}
    A::Matrix{Float64}
end
Base.size(v::MyVec) = size(v.A)
Base.getindex(v::MyVec, i::Int) = v.A[i] 
Base.size(v::MyMat) = size(v.A)
Base.getindex(v::MyMat, i::Int, j::Int) = v.A[i, j]
D = Diagonal(MyVec(rand(10)))
A = UnitUpperTriangular(MyMat(rand(10, 10)))
D * A
```
```julia-repl
julia> D * A
ERROR: MethodError: *(::Diagonal{Float64, MyVec}, ::UnitUpperTriangular{Float64, MyMat}) is ambiguous.

Candidates:
  *(D::Diagonal, A::UnitUpperTriangular)
    @ LinearAlgebra C:\Users\djv23\.julia\juliaup\julia-1.10.4+0.x64.w64.mingw32\share\julia\stdlib\v1.10\LinearAlgebra\src\diagonal.jl:578
  *(A::Diagonal{<:Any, <:LayoutVector}, B::AbstractMatrix)
    @ ArrayLayouts C:\Users\djv23\.julia\packages\ArrayLayouts\VzDAX\src\ArrayLayouts.jl:220
  *(D::Diagonal, A::LinearAlgebra.AbstractTriangular)
    @ LinearAlgebra C:\Users\djv23\.julia\juliaup\julia-1.10.4+0.x64.w64.mingw32\share\julia\stdlib\v1.10\LinearAlgebra\src\diagonal.jl:835
  *(A::AbstractMatrix, B::LinearAlgebra.AbstractTriangular)
    @ LinearAlgebra C:\Users\djv23\.julia\juliaup\julia-1.10.4+0.x64.w64.mingw32\share\julia\stdlib\v1.10\LinearAlgebra\src\triangular.jl:1500
  *(D::Diagonal, A::AbstractMatrix)
    @ LinearAlgebra C:\Users\djv23\.julia\juliaup\julia-1.10.4+0.x64.w64.mingw32\share\julia\stdlib\v1.10\LinearAlgebra\src\diagonal.jl:292

Possible fix, define
  *(::Diagonal{T, <:LayoutVector{T}} where T, ::UnitUpperTriangular)

Stacktrace:
 [1] top-level scope
   @ Untitled-1:14
```

Reply:
Is there anything else you want for this to be merged @jishnub? I'm not sure what would be defined on LinearAlgebra.jl's side but I think this would be good to go for now.